peer_gater.go

```diff
@@ -0,0 +1,302 @@
+		} else {
+			ps.tracer = &pubsubTracer{
+				internal: []internalTracer{gs.gate},
+				pid:      ps.host.ID(),
+				msgID:    ps.msgID,
+			}
+		}
+
+		return nil
+	}
+}
+
+func newPeerGater(ctx context.Context, host host.Host, threshold, decay float64) *peerGater {
+	pg := &peerGater{
+		threshold: threshold,
+		decay:     decay,
+		peerStats: make(map[peer.ID]*peerGaterStats),
+		ipStats:   make(map[string]*peerGaterStats),
+		host:      host,
+	}
+	go pg.background(ctx)
+	return pg
+}
+
+func (pg *peerGater) background(ctx context.Context) {
+	tick := time.NewTicker(DefaultDecayInterval)
+
+	defer tick.Stop()
+
+	for {
+		select {
+		case <-tick.C:
+			pg.decayStats()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (pg *peerGater) decayStats() {
+	pg.Lock()
+	defer pg.Unlock()
+
+	pg.validate *= pg.decay
+	if pg.validate < DefaultDecayToZero {
+		pg.validate = 0
+	}
+
+	pg.throttle *= pg.throttle
+	if pg.throttle < DefaultDecayToZero {
+		pg.throttle = 0
+	}
+
+	now := time.Now()
+	for ip, st := range pg.ipStats {
+		if st.connected > 0 {
+			st.deliver *= pg.decay
+			if st.deliver < DefaultDecayToZero {
+				st.deliver = 0
+			}
+
+			st.duplicate *= pg.decay
+			if st.duplicate < DefaultDecayToZero {
+				st.duplicate = 0
+			}
+
+			st.ignore *= pg.decay
+			if st.ignore < DefaultDecayToZero {
+				st.ignore = 0
+			}
+
+			st.reject *= pg.decay
+			if st.reject < DefaultDecayToZero {
+				st.reject = 0
+			}
+		} else if st.expire.Before(now) {
+			delete(pg.ipStats, ip)
+		}
+	}
+}
+
+func (pg *peerGater) getPeerStats(p peer.ID) *peerGaterStats {
+	st, ok := pg.peerStats[p]
+	if !ok {
+		st = pg.getIPStats(p)
+		pg.peerStats[p] = st
+	}
+	return st
+}
+
+func (pg *peerGater) getIPStats(p peer.ID) *peerGaterStats {
+	ip := pg.getPeerIP(p)
+	st, ok := pg.ipStats[ip]
+	if !ok {
+		st = &peerGaterStats{}
+		pg.ipStats[ip] = st
+	}
+	return st
+}
+
+func (pg *peerGater) getPeerIP(p peer.ID) string {
+	if pg.getIP != nil {
+		return pg.getIP(p)
+	}
+
+	connToIP := func(c network.Conn) string {
+		remote := c.RemoteMultiaddr()
+		ip, err := manet.ToIP(remote)
+		if err != nil {
+			return "<unknown>"
+		}
+		return ip.String()
+	}
+
+	conns := pg.host.Network().ConnsToPeer(p)
+	switch len(conns) {
+	case 0:
+		return "<unknown>"
+	case 1:
+		return connToIP(conns[0])
+	default:
+		// we have multiple connections -- order by number of streams and use the one with the
+		// most streams; it's a nightmare to track multiple IPs per peer, so pick the best one.
+		streams := make(map[string]int)
+		for _, c := range conns {
+			streams[c.ID()] = len(c.GetStreams())
+		}
+		sort.Slice(conns, func(i, j int) bool {
+			return streams[conns[i].ID()] > streams[conns[j].ID()]
+		})
+		return connToIP(conns[0])
+	}
+}
+
+func (pg *peerGater) AcceptFrom(p peer.ID) AcceptStatus {
+	if pg == nil {
+		return AcceptAll
+	}
+
+	pg.Lock()
+	defer pg.Unlock()
+
+	if time.Since(pg.lastThrottle) > PeerGaterQuiet {
+		return AcceptAll
+	}
```

Comment:
So this is a circuit breaker construction, right? At this point you close the circuit again, and if the peer misbehaves, they will be throttled again?

Reply:
right, if there hasn't been any queue throttling for 1min, we open it up again.

Reply:
if spam reappears and the queue starts throttling again, the circuit breaker will turn back on.